src/main/java/org/opensearch/knn/plugin/BlockedModelSet.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.plugin;
+
+import org.apache.logging.log4j.LogManager;
+import org.opensearch.Version;
+import org.opensearch.cluster.Diff;
+import org.opensearch.cluster.NamedDiff;
+import org.opensearch.cluster.metadata.Metadata;
+import org.opensearch.common.io.stream.StreamInput;
+import org.opensearch.common.io.stream.StreamOutput;
+import org.opensearch.common.xcontent.XContentBuilder;
+import org.opensearch.common.xcontent.XContentParser;
+import org.apache.logging.log4j.Logger;
+
+import java.io.IOException;
+import java.util.EnumSet;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+// This class implements Metadata.Custom Interface to store a set of modelIds in the cluster metadata.
+// The modelIds of the models that are under deletion are added to this set and later removed from this set after deletion.
+// Also, this class implements the methods to perform operations on this set (like add, remove, contains)
+public class BlockedModelSet implements Metadata.Custom {
+
+    public static Logger logger = LogManager.getLogger(BlockedModelSet.class);
+    public static final String TYPE = "opensearch-knn-blocked-models";
+    private Set<String> blockedModelSet;
+
+    public BlockedModelSet(Set<String> blockedModelSet) {
+        this.blockedModelSet = blockedModelSet;
+    }
+
+    public BlockedModelSet() {
+        this.blockedModelSet = new HashSet<>();
+    }
+
+    @Override
+    public EnumSet<Metadata.XContentContext> context() {
+        return Metadata.ALL_CONTEXTS;
+    }
+
+    @Override
+    public Diff<Metadata.Custom> diff(Metadata.Custom custom) {
+        return null;
+    }
+
+    @Override
+    public String getWriteableName() {
+        return TYPE;
+    }
+
+    @Override
+    public Version getMinimalSupportedVersion() {
+        return Version.CURRENT.minimumCompatibilityVersion();
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeStringCollection(blockedModelSet);
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        return builder;
+    }
+
+    public BlockedModelSet(StreamInput in) throws IOException {
+        this.blockedModelSet = new HashSet<>(in.readStringList());
+    }
+
+    public void remove(String modelId) {
+        blockedModelSet.remove(modelId);
```

Comment:
Im wondering if this class needs to be read only. I see we get it from cluster metadata in the update action. But would updating it here change the cluster state on this node before the leader node accepts the update?